optuna/samplers/_tpe/multivariate_parzen_estimator.py

```diff
@@ -0,0 +1,371 @@
+        self._categorical_weights = {}  # type: Dict[str, Optional[np.ndarray]]
+        for param_name, dist in search_space.items():
+            observations = multivariate_observations[param_name]
+            if isinstance(dist, distributions.CategoricalDistribution):
+                mus = sigmas = None
+                categorical_weights = self._calculate_categorical_params(observations, param_name)
+            else:
+                mus, sigmas = self._calculate_parzen_est_params(observations, param_name)
+                categorical_weights = None
+            self._mus[param_name] = mus
+            self._sigmas[param_name] = sigmas
+            self._categorical_weights[param_name] = categorical_weights
+
+    def sample(self, rng: np.random.RandomState, size: int) -> Dict[str, np.ndarray]:
+
+        multivariate_samples = {}
+        active = rng.choice(len(self._weights), size, p=self._weights)
+
+        for param_name, dist in self._search_space.items():
+
+            if isinstance(dist, distributions.CategoricalDistribution):
+                categorical_weights = self._categorical_weights[param_name]
+                assert categorical_weights is not None
+                weights = categorical_weights[active, :]
+                samples = _MultivariateParzenEstimator._sample_from_categorical_dist(rng, weights)
+
+            else:
+                # We restore parameters of parzen estimators.
+                low = self._low[param_name]
+                high = self._high[param_name]
+                mus = self._mus[param_name]
+                sigmas = self._sigmas[param_name]
+                assert low is not None
+                assert high is not None
+                assert mus is not None
+                assert sigmas is not None
+
+                # We sample from truncnorm.
+                trunc_low = (low - mus[active]) / sigmas[active]
+                trunc_high = (high - mus[active]) / sigmas[active]
+                samples = np.full((), fill_value=high + 1.0, dtype=np.float64)
+                while (samples >= high).any():
+                    samples = np.where(
+                        samples < high,
+                        samples,
+                        truncnorm.rvs(
+                            trunc_low,
+                            trunc_high,
+                            size=size,
+                            loc=mus[active],
+                            scale=sigmas[active],
+                            random_state=rng,
+                        ),
+                    )
+            multivariate_samples[param_name] = samples
+        multivariate_samples = self._transform_from_uniform(multivariate_samples)
+        return multivariate_samples
+
+    def log_pdf(self, multivariate_samples: Dict[str, np.ndarray]) -> np.ndarray:
+
+        multivariate_samples = self._transform_to_uniform(multivariate_samples)
+        n_weights = len(self._weights)
```

Comment:
This is equal to the number of observations? How about using `n_observations`?
```suggestion
        n_observations = len(self._weights)
```

Reply:
Yes, it is. Still, I am a little afraid similarity of words samples and observations might cause confusion. On the other hand, `n_weights` and `n_observations` being the same might cause another confusion as you pointed out.

Reply:
Probably, changing all `n_observations` to `n_weights` is a good idea

Reply:
Hmm. I think `n_weights` is misleading. When I see this, I wonder what weight is for `n_weight`. I think we already have the naming convention of `sample` and `observation`. Therefore, I think `n_observations` is reasonable here.